src/pint/models/model_builder.py

```diff
@@ -26,276 +26,296 @@ class UnknownBinaryModel(ValueError):
-                # remove this bogus alias
-                try:
-                    ix = self.timing_model[p].aliases.index(self.timing_model[p].prefix)
-                except ValueError:
-                    pass
-                else:
-                    del self.timing_model[p].aliases[ix]
-        if parfile is not None:
-            self.timing_model.read_parfile(parfile)
-
-    def get_control_info(self):
-        info = {}
-        if not self.param_unrecognized == {}:
-            for ctrlp in self.control_params:
-                if ctrlp in self.param_unrecognized:
-                    info[ctrlp] = self.param_unrecognized[ctrlp]
+        result_componets = []
+        param_inpar = self.parse_parfile(parfile)
+        # 1. iteration read parfile with a no component timing_model to get
+        # the overall control parameters. This will get us the binary model name
+        # build the base fo the timing model
+        self.timing_model.read_parfile(parfile)
+        # Get the binary model name
+        binary = self.timing_model.BINARY.value
+        if binary is not None:
+            result_componets.append(self.search_pulsar_system_components(binary))
+        # 2. Get the component list from the parameters in the parfile.
+        # 2.1 Check the aliases of input parameters.
+        # This does not include the repeating parameters, but it should not
+        # matter in the component selection.
+        unrec_param = [] # For Unrecognized parameters.
+        pint_par = []
+        for pp in param_inpar:
+            p_name = self.alias_2_pint_param(pp)
+            if p_name is not None:
+                pint_par.append(p_name)
+            else:
+                unrec_param.append(pp)
+
+        param_components_inpar = {p: self.param_component_map[p] for p in pint_par}
+        # Back map the possible_components and the parameters in the parfile
+        # This will remove the duplicate components.
+        possible_components = {}
+        for k, cps in param_components_inpar.items():
+            for cp in cps:
+                if cp == 'timing_model': # skip the timing model parameters.
+                    continue
+                if cp in possible_components.keys():
+                    possible_components[cp].append(k)
                 else:
-                    # Check if the prefix match
-                    for p in self.control_params.keys():
-                        if p.startswith(ctrlp):
-                            info[ctrlp] = self.param_unrecognized[ctrlp]
-        return info
+                    possible_components[cp] = [k]
+        # Get conflicting components based on categories
+        # conflict_components = {category: [component1, component2, ...]}
+        conflict_components = {}
+        for k in possible_components.keys():
+            cate = self.component_category_map[k]
+            if cate in conflict_components.keys():
```

Comment:
`collections.defaultdict(list)` does this for you.

Reply:
Done